File: app/src/main/java/com/manoj/dlt/features/ProfileFeature.java
package com.manoj.dlt.features;

import android.content.Context;
import android.util.Log;

import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.manoj.dlt.Constants;
import com.manoj.dlt.DbConstants;
import com.manoj.dlt.interfaces.IProfileFeature;
import com.manoj.dlt.utils.Utilities;

import java.util.UUID;

public class ProfileFeature implements IProfileFeature
{
    private static ProfileFeature _instance;
    private FileSystem _fileSystem;
    private String _userId;

    private ProfileFeature(Context context)
    {
        _fileSystem = Utilities.getOneTimeStore(context);
        _userId = _fileSystem.read(Constants.USER_ID_KEY);
        if(_userId == null)
        {
            _userId = generateUserId();
            _fileSystem.write(Constants.USER_ID_KEY, _userId);
        }
        Log.d("profile","user id = "+_userId);
    }

    public static ProfileFeature getInstance(Context context)
    {
        if(_instance == null)
        {
            _instance = new ProfileFeature(context);
        }
        return _instance;
    }

    @Override
    public String getUserId()
    {
        return _userId;
    }

    @Override
    public DatabaseReference getCurrentUserFirebaseBaseRef()
    {
        DatabaseReference baseUserRef = Constants.getFirebaseUserRef();
        return baseUserRef.child(_userId);
    }

    private String generateUserId()
    {
        //TODO: better implementation
        String rand = UUID.randomUUID().toString();
        return rand.substring(0,5);
    }
}

File: app/src/main/kotlin/com/manoj/dlt/features/ProfileFeature.kt
package com.manoj.dlt.features

import android.content.Context
import android.util.Log
import com.google.firebase.database.DatabaseReference
import com.manoj.dlt.Constants
import com.manoj.dlt.interfaces.IProfileFeature
import com.manoj.dlt.utils.Utilities
import java.util.*

class ProfileFeature private constructor(context: Context) : IProfileFeature {
    private val fileSystem = Utilities.getOneTimeStore(context)

    override var userId = fileSystem.read(Constants.USER_ID_KEY) ?: generateUserIdAndSave()

    override val currentUserFirebaseBaseRef: DatabaseReference
        get() = Constants.firebaseUserRef.child(userId)

    private fun generateUserIdAndSave(): String {
        val generatedUserId = generateUserId()
        fileSystem.write(Constants.USER_ID_KEY, generatedUserId)
        Log.d("profile", "user id = " + generatedUserId)
        return generatedUserId
    }

    private fun generateUserId(): String {
        //TODO: better implementation
        val rand = UUID.randomUUID().toString()
        return rand.substring(0, 5)
    }

    companion object {
        private var instance: ProfileFeature? = null

        fun getInstance(context: Context): ProfileFeature {
            if (instance == null) {
                instance = ProfileFeature(context)
            }
            return instance as ProfileFeature
        }
    }
}
